Mobile: Upgrade to React Native 0.79

personalizedrefrigerator opened the pull request:
# Summary

Updates to React Native 0.79 and react-native-web 0.20.

This also:
- Requires an upgrade to React 19.
- Upgrades `@testing-library/react-native` to v13.2.0 for React 19 support.

# Notes

- This may fix an issue related to focus management and external keyboards on Android.

# To-do

- [ ] Testing plan & more in-depth manual testing.
    - [ ] Test alarm notifications.
      - [x] iOS 18 (Simulator)
      - [ ] Android
    - [ ] Test quick actions
      - [x] iOS 18
      - [ ] Android
    - [ ] Test sync.
        - [ ] File System (Android)
        - [ ] Dropbox (iOS)
    - [ ] Test biometrics.
        - [ ] iOS 18 (physical device)
        - [ ] Android
    - [ ] Test editing/viewing a note with an audio attachment.
      - [ ] iOS 18 (Simulator)
      - [ ] Android
    - [ ] Test attaching files.
      - [ ] iOS 18 (Simulator)
      - [ ] Android
    - [ ] Test taking photos
       - [ ] iOS 18 (physical device)
       - [ ] Android 13
    - [ ] Test recording audio
       - [ ] iOS 18 (physical device)
       - [ ] Android 13
    - [ ] Test JEX export & import
        - [ ] iOS 18 (Simulator), single notebook with 2-3 notes.
        - [ ] Android 13, 3-4 notebooks with several total notes.
- [ ] Further explanation of changes to `screens/Note/Note.tsx` (passing `state` to `scheduleSave` to fix a save-related race condition observed in tests).

<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/readme/dev/index.md

-->

personalizedrefrigerator commented on packages/app-mobile/components/EditorToolbar/EditorToolbar.test.tsx:
```diff
@@ -46,11 +45,19 @@ const toggleSettingsItem = async (props: ToggleSettingItemProps) => {
 	expect(itemCheckbox).toBeVisible();
-	expect(itemCheckbox).toHaveAccessibilityState({ checked: initialChecked });
+	if (initialChecked) {
+		expect(itemCheckbox).toBeChecked();
+	} else {
+		expect(itemCheckbox).not.toBeChecked();
+	}
```
The new `@testing-library/react-native` version removed [*AccessibilityState queries](https://callstack.github.io/react-native-testing-library/docs/migration/v13#removed-deprecated-byaccessibilitystate-queries).

personalizedrefrigerator commented on packages/app-mobile/components/screens/ConfigScreen/plugins/PluginStates.search.test.tsx:
```diff
@@ -2,7 +2,6 @@ import * as React from 'react';
 import { mockMobilePlatform, setupDatabaseAndSynchronizer, switchClient } from '@joplin/lib/testing/test-utils';
 
 import { render, screen, userEvent, waitFor } from '@testing-library/react-native';
-import '@testing-library/react-native/extend-expect';
```
Jest matchers are [now extended by default and `extend-expect` imports can be removed](https://callstack.github.io/react-native-testing-library/docs/migration/v13#extend-jest-matchers-by-default).

personalizedrefrigerator commented:
Issue observed during testing: On iOS, the camera doesn't work the first time it's started, just after an "allow permission" dialog is shown.

personalizedrefrigerator commented on .yarn/patches/react-native-npm-0.79.2-9db13eddfe.patch:
```diff
@@ -0,0 +1,205 @@
+# This patch fixes two issues:
+# - Updates RCTDeviceInfo.m to match https://github.com/facebook/react-native/commit/0b8db7e5e814cfbf9974cc5b6ceb64e8006d8a3c.
+#   This fixes an issue in which useWindowDimensions returns incorrect
+#   values in landscape mode in iOS.
+#   This should be fixed in React Native 0.80. See https://github.com/facebook/react-native/issues/51086.
```
Alternative: Wait for RN 0.80 to be released.

laurent22 replied:
I think we can merge this one already since you've already done the work on it, and it will make the upgrade to 0.80 easier. Is there anything else which means that this PR is in draft status?

(also there's a conflict in package.json)

personalizedrefrigerator replied:
> Is there anything else which means that this PR is in draft status?

There's a camera-related issue that could have:
- Been caused by this pull request,
- Been caused by the upgrade to 0.77,
- Existed before the upgrade to 0.77.

In particular, the camera needs to be closed and re-opened on iOS after the user gives the app camera permission for the first time. A potential workaround might be to not load the `expo-camera` view until the app has been granted camera permissions.

personalizedrefrigerator commented:
**Note**: CI is failing because tests for `components/screens/ShareNoteDialog.test.tsx` need to be updated for the new react-native-testing-library version.

laurent22 commented:
Sorry there are more conflicts as a few Renovate updates have been merged. Feel free to merge the PR after fixing the conflicts as there may be more Renovate PRs

personalizedrefrigerator commented:
I've made a few changes since the last review:
- Removed debugging code that should not have been committed in https://github.com/laurent22/joplin/pull/12337/commits/8c8be495adf7597301209300e8b70734bcd23fb2.
    - One instance of a bundle ID change — among the changes I regularly apply locally to run on a physical iOS device — was committed.
    - `.debug` logging was included that isn't directly related to this PR.
- Updated `gradle-wrapper.jar` in https://github.com/laurent22/joplin/pull/12337/commits/4b9e852b54c4986f7a80a597deb815b5c2104418.
    - The `sha256sum` of the file is 81a82aaea5abcc8ff68b3dfcb58b3c3c429378efd98e7433460610fecd7ae45f, which matches the "wrapper JAR checksum" for v8.13 posted on https://gradle.org/release-checksums/.
- Additional commenting in https://github.com/laurent22/joplin/pull/12337/commits/f86c595ee4a3dd29acaac16858956020db7704b9.
- Fixed an issue that prevented the camera from being used on first launch on iOS in  https://github.com/laurent22/joplin/pull/12337/commits/5b87e9488b5b21b4bdddf74cc23989e92eb14a7d.
